src/high_level/kex.rs

```diff
@@ -0,0 +1,158 @@
+    }
+}
+
+#[derive(Debug, PartialEq)]
+/// A key pair used to establish shared keys for a single session.
+pub struct EphemeralServerSession {
+    private_key: x25519::PrivateKey,
+    public_key: PublicKey,
+}
+
+impl EphemeralServerSession {
+    /// Generate a new random key pair.
+    pub fn new() -> Result<Self, UnknownCryptoError> {
+        let privkey = x25519::PrivateKey::generate();
+        let pubkey: PublicKey = PublicKey::try_from(&privkey)?;
+
+        Ok(Self {
+            private_key: privkey,
+            public_key: pubkey,
+        })
+    }
+
+    /// Get a reference to the public key of the key pair.
+    pub fn get_public(&self) -> &PublicKey {
+        &self.public_key
+    }
+
+    /// Establish session keys with a client. This moves `self` to ensure that the keys
+    /// generated with [`Self::new()`] are only used for this key exchange, thus remaining ephemeral.
+    pub fn establish_with_client(
+        self,
+        client_public_key: &PublicKey,
+    ) -> Result<SessionKeys, UnknownCryptoError> {
+        let q = x25519::key_agreement(&self.private_key, client_public_key)?;
+
+        establish_session_keys(&q, client_public_key, &self.public_key)
+    }
+}
+
+#[derive(Debug, PartialEq)]
+/// A set of shared secrets for either transmitting to this entity or send to another party.
+pub struct SessionKeys {
+    rx: SharedKey,
+    tx: SharedKey,
+}
+
+impl SessionKeys {
+    /// Get the shared secret intended to be used for receiving data from the other party.
+    pub fn get_receiving(&self) -> &SharedKey {
+        &self.rx
+    }
+
+    /// Get the shared secret intended to be used for transporting data to the other party.
+    pub fn get_transport(&self) -> &SharedKey {
+        &self.tx
+    }
+}
+
+/// Using BLAKE2b, derive two shared secret from a scalarmult computation.
+fn establish_session_keys(
+    shared_secret: &SharedKey,
+    client_pk: &PublicKey,
+    server_pk: &PublicKey,
+) -> Result<SessionKeys, UnknownCryptoError> {
+    let mut ctx = Blake2b::new(None, 64)?;
+    ctx.update(shared_secret.unprotected_as_bytes())?;
+    ctx.update(client_pk.as_ref())?;
+    ctx.update(server_pk.as_ref())?;
+    let keys = ctx.finalize()?;
+
+    Ok(SessionKeys {
+        rx: SharedKey::from_slice(&keys.as_ref()[..32])?,
+        tx: SharedKey::from_slice(&keys.as_ref()[32..])?,
+    })
+}
+
+#[test]
+fn test_ephemeral() {
+    let session_server = EphemeralServerSession::new().unwrap();
+    let server_public_key = session_server.get_public().clone();
+
+    let session_client = EphemeralClientSession::new().unwrap();
+    let client_public_key = session_client.get_public().clone();
+
+    let client = session_client
+        .establish_with_server(&server_public_key)
+        .unwrap();
+    let server = session_server
+        .establish_with_client(&client_public_key)
+        .unwrap();
+
+    assert_eq!(client, server);
+    assert_eq!(client.get_receiving(), server.get_receiving());
```

Comment:
This is of course not correct. Should be `Client.RX == Server.TX` and `Client.TX == Server.RX`